src/nv_ingest/modules/transforms/text_splitter.py

```diff
@@ -0,0 +1,175 @@
+        chunk_offsets = offsets[: len(chunk)]
+        start_offset = chunk_offsets[0][0]
+        end_offset = chunk_offsets[-1][1]
+
+        # Extract the original text for this chunk based on offsets
+        text_chunk = text[start_offset:end_offset]
+        text_chunks.append(text_chunk)
+
+        # Remove processed offsets for the next iteration
+        offsets = offsets[len(chunk) :]
+
+    return text_chunks
+
+
+MODULE_NAME = "text_splitter"
+MODULE_NAMESPACE = "nv_ingest"
+
+TextSplitterLoaderFactory = ModuleLoaderFactory(MODULE_NAME, MODULE_NAMESPACE, TextSplitterSchema)
+
+
+@register_module(MODULE_NAME, MODULE_NAMESPACE)
+def _text_splitter(builder: mrc.Builder):
+    """
+    A pipeline module that splits documents into smaller parts based on the specified criteria.
+    """
+
+    validated_config = fetch_and_validate_module_config(builder, TextSplitterSchema)
+
+    @filter_by_task(["split"])
+    @traceable(MODULE_NAME)
+    @cm_skip_processing_if_failed
+    @nv_ingest_node_failure_context_manager(
+        annotation_id=MODULE_NAME,
+        raise_on_failure=validated_config.raise_on_failure,
+    )
+    def split_and_forward(message: ControlMessage):
+        try:
+            # Assume that df is going to have a 'content' column
+            task_props = message.remove_task("split")
+
+            if isinstance(task_props, BaseModel):
+                task_props = task_props.model_dump()
+
+            # Validate that all 'content' values are not None
+            with message.payload().mutable_dataframe() as mdf:
+                df = mdf.to_pandas()
+
+            # Filter to document type
+            bool_index = df["document_type"] == ContentTypeEnum.TEXT
+            df_filtered = df.loc[bool_index]
+
+            if df_filtered.empty:
+                return message
+
+            # Override parameters if set
+            tokenizer = task_props.get("tokenizer", validated_config.tokenizer)
+            chunk_size = task_props.get("chunk_size", validated_config.chunk_size)
+            chunk_overlap = task_props.get("chunk_overlap", validated_config.chunk_overlap)
+            params = task_props.get("params", {})
+
+            hf_access_token = params.get("hf_access_token", None)
+            split_source_types = params.get("split_source_types", ["TEXT"])
+
+            logger.debug(
+                f"Splitting text with tokenizer: {tokenizer}, "
+                f"chunk_size: {chunk_size} tokens, "
+                f"chunk_overlap: {chunk_overlap}"
+            )
+
+            # Filter to file type
+            bool_index = pd.json_normalize(df_filtered["metadata"])["source_metadata.source_type"].isin(
+                split_source_types
+            )
+            df_filtered = df_filtered.loc[bool_index]
+
+            if df_filtered.empty:
+                return message
+
+            tokenizer_model = AutoTokenizer.from_pretrained(tokenizer, token=hf_access_token)
+
+            split_docs = []
```

Comment:
Matching documents should be collected and processed all at once as opposed to iterating rows; and we should actually make this a multiprocessing stage so we're able to use the worker pool for CPU bound tasks. But we can hold of if this needs to go in sooner.